### PR TITLE
change args parser and add a set_dns function

### PR DESCRIPTION
**Pull request by attenuation**
1.  use rust-2018 version grammar
2. change getopts to the more modern clap
3. add dns information to Struct Response, then we can specify a dns server to client.

Then I will add KCP support to kytan 😁

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -48,15 +48,27 @@ $ sudo iptables -t nat -A POSTROUTING -s 10.10.10.0/24 -o <INTERFACE> -j MASQUER
 To run `kytan` in server mode and listen on UDP port `9527` with password `hello`:
 
 ```
-$ sudo ./kytan -m s -p 9527 -s hello
+
+$ sudo ./kytan server -k hello 
+```
+if you want open log display(`info` is log level,you can change it by your idea)
```
Comment (changlan) [inline README.md]: Minor fix (space and letter case) ->
If you want open log display (`info` is log level, you can change it by your idea)

File: src/cli.rs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+            port: port,
+            key: key_str.to_string(),
+            dns: dns
+        }))
+    } else {
+        unimplemented!()
+    }
+}
+
+// pub fn get_args() -> Result<Args,String> {
```
Comment (changlan) [inline src/cli.rs]: How about removing the commented example?

File: src/main.rs [hunk 3]
```diff
@@ -53,40 +40,45 @@ fn main() {
         panic!("Please run as root");
     }
 
-    let mut opts = getopts::Options::new();
-    opts.reqopt("m", "mode", "mode (server or client)", "[s|c]");
-    opts.optopt("p", "port", "UDP port to listen/connect", "PORT");
-    opts.optopt("h", "host", "remote host to connect (client mode)", "HOST");
-    opts.optopt("s", "secret", "shared secret", "PASSWORD");
+    // let mut opts = getopts::Options::new();
```
Comment (changlan) [inline src/main.rs]: Same here

File: src/main.rs [hunk 4]
```diff
@@ -53,40 +40,45 @@ fn main() {
...
-            network::connect(&host, port, true, &secret)
-        }
-        _ => unreachable!(),
-    };
+    match cli::get_args().unwrap() {
+        cli::Args::Client(client) => network::connect(&client.remote_addr, client.port, client.default_route, &client.key),
+        cli::Args::Server(server) => network::serve(server.port, &server.key, server.dns),
+    }
+
+    // match args.mode.as_ref() {
```
Comment (changlan) [inline src/main.rs]: Same here

File: src/network.rs [hunk 5]
```diff
@@ -84,33 +87,33 @@ fn derive_keys(password: &str) -> (aead::SealingKey, aead::OpeningKey) {
...
     assert_eq!(&recv_addr, addr);
     info!("Response received from {}.", addr);
     let (aad,nonce) = generate_add_nonce(secret);
     let decrypted_buf = aead::open_in_place(&opening_key, nonce, aad, 0, &mut buf[0..len]).unwrap();
     let dlen = decrypted_buf.len();
-    let resp_msg: Message = try!(deserialize(&decrypted_buf[0..dlen]).map_err(|e| e.to_string()));
+    let resp_msg: Message = deserialize(&decrypted_buf[0..dlen]).map_err(|e| e.to_string())?;
     match resp_msg {
-        Message::Response { id, token } => Ok((id, token)),
+        Message::Response { id, token,dns } => Ok((id, token,dns)),
```
Comment (changlan) [inline src/network.rs]: space after comma

File: src/network.rs [hunk 6]
```diff
@@ -227,7 +230,7 @@ pub fn connect(host: &str, port: u16, default: bool, secret: &str) {
     }
 }
 
-pub fn serve(port: u16, secret: &str) {
+pub fn serve(port: u16, secret: &str,dns: IpAddr) {
```
Comment (changlan) [inline src/network.rs]: space after comma

File: src/network.rs [hunk 7]
```diff
@@ -321,7 +325,7 @@ pub fn serve(port: u16, secret: &str) {
                                         .unwrap();
                             }
                         }
-                        Message::Response { id: _, token: _ } => {
+                        Message::Response { id: _, token: _ ,dns: _} => {
```
Comment (changlan) [inline src/network.rs]: space after comma

File: src/network.rs [hunk 8]
```diff
@@ -406,7 +410,7 @@ mod tests {
     #[cfg(target_os = "linux")]
     fn integration_test() {
         assert!(utils::is_root());
-        let server = thread::spawn(move || serve(8964, "password"));
+        let server = thread::spawn(move || serve(8964, "password","8.8.8.8".parse::<IpAddr>().unwrap()));
```
Comment (changlan) [inline src/network.rs]: space after comma

File: src/network.rs [hunk 9]
```diff
@@ -415,7 +419,7 @@ mod tests {
         let local_addr: SocketAddr = "0.0.0.0:0".parse::<SocketAddr>().unwrap();
         let local_socket = UdpSocket::bind(&local_addr).unwrap();
 
-        let (id, token) = initiate(&local_socket, &remote_addr, "password").unwrap();
+        let (id, token,dbs) = initiate(&local_socket, &remote_addr, "password").unwrap();
```
Comment (changlan) [inline src/network.rs]: space after comma

File: src/utils.rs [hunk 10]
```diff
@@ -53,26 +50,32 @@ pub enum RouteType {
 pub struct DefaultGateway {
     origin: String,
     remote: String,
+    default: bool,
 }
 
 impl DefaultGateway {
-    pub fn create(gateway: &str, remote: &str) -> DefaultGateway {
+    pub fn create(gateway: &str, remote: &str,default: bool) -> DefaultGateway {
```
Comment (changlan) [inline src/utils.rs]: space after comma

File: src/utils.rs [hunk 11]
```diff
@@ -53,26 +50,32 @@ pub enum RouteType {
...
         let origin = get_default_gateway().unwrap();
         info!("Original default gateway: {}.", origin);
         add_route(RouteType::Host, remote, &origin).unwrap();
-        delete_default_gateway().unwrap();
-        set_default_gateway(gateway).unwrap();
+        if default{
```
Comment (changlan) [inline src/utils.rs]: space before `{`

File: src/utils.rs [hunk 12]
```diff
@@ -184,21 +187,66 @@ pub fn get_public_ip() -> Result<String, String> {
...
         delete_route(RouteType::Host, "1.1.1.1").unwrap();
+        assert!(!get_route_gateway("1.1.1.1").unwrap().contains(&*gw));
     }
+    
+    #[test]
+    fn set_dns_test() {
+        assert!(is_root());
+        set_dns("8.8.8.8").unwrap();
+    }
+
```
Comment (changlan) [inline src/utils.rs]: blank line

**Comment by changlan** [comment]
Thanks! This PR has been merged.

**Comment by attenuation** [comment]
Can you try to solve my issues?

**Pull request merged**